File: ios/CropView.m
//
//  CropInterface.m
//  LiveInk
//
//  Created by Swarup Mahanti on 8/23/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import "CropView.h"
#import "CropControl.h"
#import "CropRectangle.h"
#import "Globals.h"

const CGFloat MARGIN = 100;

@interface CropView(){
    NSMutableArray<CropControl*> *_controlIcons;
    CropControl *_touchedIcon;
    CropRectangle *_cropRect;
    CGRect _cropBox;
    BOOL _needsLayout;
    CGPoint previousLoc;
}
@end

@implementation CropView

-(instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        _controlIcons = [NSMutableArray array];
        _cropBox = CGRectZero;
        _needsLayout = true;
    }
    return self;
}

-(void)layoutSubviews {
    [super layoutSubviews];
    self.backgroundColor = [[UIColor alloc] initWithWhite:0.1 alpha:0.5];
    if (_needsLayout == YES) {
        _cropBox = CGRectInset(self.bounds, MARGIN, MARGIN);
        [self createControlIcons];
        _needsLayout = NO;
    }
}

-(void)createControlIcons {
    if ([_controlIcons count] > 0) {
        for (CropControl *icon in _controlIcons) {
            [icon removeFromSuperview];
        }
        [_controlIcons removeAllObjects];
    }
    for (int i = 0; i < 4; i++) {
        CropControl *icon = [[CropControl alloc] initWithFrame:_cropBox index:i];
        //NSDLog(@"index:%d frame:%@", i, NSStringFromCGRect(icon.frame));
        icon.userInteractionEnabled = YES;
        [_controlIcons addObject:icon];
        [self addSubview:icon];
    }
}

-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {

    if ([touches anyObject] == nil) {
        return;
    }
    UITouch *touch = [[event allTouches] anyObject];
    CGPoint touchPoint = [touch locationInView:self];
    _touchedIcon = [self iconTouched:touchPoint];
    _cropRect = [[CropRectangle alloc] initWith:_cropBox];
    if (_touchedIcon) {
        [_cropRect setPivotIndex:_touchedIcon.index];
        [_touchedIcon activate:YES resize:YES];
    }
    previousLoc = touchPoint;
}

-(void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {

    UITouch *touch = [[event allTouches] anyObject];
    CGPoint current = [touch locationInView:self];
    CGPoint previous = [touch previousLocationInView:self];
    CGFloat dx = current.x - previousLoc.x;
    CGFloat dy = current.y - previousLoc.y;
    //[_cropRect setPivotLocation:current];
    [_cropRect touchDragged:dx :dy];
    _cropBox = [_cropRect rectRegion];
    [self updateControlPoints];

    previousLoc = current;

    [self setNeedsDisplay];
}

-(void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (_touchedIcon) {
        [_touchedIcon deActivate];
    }

  NSDictionary *box = @{
    @"rect": @[
        @(_cropBox.origin.x),
        @(_cropBox.origin.y),
        @(_cropBox.size.width),
        @(_cropBox.size.height)
    ]
  };

  _onCropBoxChange(box);
}


-(void)updateControlPoints {
    for (CropControl* icon in _controlIcons) {
        icon.center = [_cropRect pointForIndex:icon.index];
    }
}

-(void)drawRect:(CGRect)rect {
    [super drawRect:rect];

    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSaveGState(context);
    CGContextClearRect(context, _cropBox);
    CGContextSetBlendMode(context, kCGBlendModeClear);
    const CGFloat* components = CGColorGetComponents([[UIColor colorWithWhite:0.0 alpha:0.0] CGColor]);
    CGContextSetFillColor(context, components);
    CGContextFillRect(context, _cropBox);
    CGContextRestoreGState(context);
    //CGContextRelease(context);
}



-(CropControl*)iconTouched:(CGPoint)point {
    for (CropControl* icon in _controlIcons) {
        if (CGRectContainsPoint(icon.frame, point)) {
            return icon;
        }
    }
    return nil;
}

-(void)sync:(CGRect)box{
}



@end

File: ios/CropView.swift
//
//  CropView.swift
//  LiveInk
//
//  Created by Swarup Mahanti on 8/1/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

import Foundation

class CropView: UIView {

  private var maskBorder: CGFloat = 1.0
  private var maskColor: UIColor = UIColor.clear {
    didSet{
      DEBUGPRINT("\(maskColor)")
      //cropControl.backgroundColor = maskColor
    }
  }
  var controlIcons = [ControlPointIcon]()
  var controlRect: ControlRect?
  var touchedIcon: ControlPointIcon?
  var cropBox: CGRect? = nil
  var touchBegin: CGPoint = CGPoint.zero
  var onCropBoxChange: RCTDirectEventBlock!{
    didSet{
      DEBUGPRINT("\(onCropBoxChange)")
    }
  }

  override init(frame: CGRect) {
    super.init(frame: frame)
  }

  required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    //self.addSubview(cropControl)
    self.backgroundColor = UIColor.init(white: 0.1, alpha: 0.5)
    if cropBox == nil {
      cropBox = self.bounds.insetBy(dx: 100, dy: 100)
      createControlIcons()
      sync(box: cropBox!)
    }
  }

  func createControlIcons() {
    for icon in controlIcons {
      icon.removeFromSuperview()
    }
    controlIcons.removeAll()
    for i in 0..<8 {
      let icon = ControlPointIcon(bbox: cropBox!, idx:i)
      icon.isUserInteractionEnabled = false
      self.addSubview(icon)
      controlIcons.append(icon)
    }
  }

  func setMaskBorderWidth(border: CGFloat) {
    maskBorder = border
  }
  func setMaskColor(color: UIColor) {
    maskColor = color
  }

  override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard let touch = touches.first else {
      return
    }
    touchBegin = touch.location(in: self)
    touchedIcon = iconTouched(point: touchBegin)
    touchedIcon?.activate(resize: true)
    let corner: Int = touchedIcon?.i ?? -1
    if corner != -1 {
      controlRect = ControlRect.init(rect: cropBox!, idx: corner)
    }
  }

  override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard let touch = touches.first else {
      return
    }
    if touchedIcon != nil {
      resize(icon: touchedIcon!, from: touchBegin, to: touch.location(in: self))
    }else {
      move(from: touch.previousLocation(in: self), to: touch.location(in: self))
      updateControlIcons()
    }
    self.setNeedsDisplay()
  }

  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
    touchedIcon?.deactivate()
    sync(box: cropBox!)
  }

  func sync(box: CGRect){
    var data: [String: Any?] = [String: Any?].init()
    data["box"] = box.toJSON()
    self.onCropBoxChange(data)
  }

  func move(from: CGPoint, to: CGPoint) {
    let dx = to.x - from.x
    let dy = to.y - from.y
    var center = cropBox!.center
    center.x += dx
    center.y += dy
    cropBox = CGRect.init(x: center.x - cropBox!.width/2, y: center.y - cropBox!.height/2, width: cropBox!.width, height: cropBox!.height)
  }

  func resize(icon:ControlPointIcon, from: CGPoint, to: CGPoint){
    cropBox = getFrame(dx: to.x - from.x, dy: to.y - from.y)
    updateControlIcons()
  }

  func updateControlIcons() {
    for icon in controlIcons {
      icon.updateFrame(bbox: cropBox!)
    }
  }

  func getFrame(dx: CGFloat, dy:CGFloat) -> CGRect {
    controlRect!.dragControlPoint(dx: dx, dy: dy)
    return controlRect!.frame
  }

  func iconTouched(point: CGPoint) -> ControlPointIcon? {
    for icon in controlIcons {
      if icon.frame.contains(point) {
        return icon
      }
    }
    return nil
  }

  override func draw(_ rect: CGRect) {
    super.draw(rect)

    let context: CGContext = UIGraphicsGetCurrentContext()!
    context.saveGState()

    context.clear(cropBox!)
    context.setBlendMode(CGBlendMode.clear)
    context.setFillColor(UIColor.init(white: 0.0, alpha: 0.0).cgColor)
    context.fill(cropBox!)
    context.restoreGState()
  }
}
